File: C/pingWithChecksum.c
#include <stdio.h>
#include <stdlib.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <fcntl.h>
#include <unistd.h>
#include <netdb.h>

int32_t checksum(uint16_t *buf, int32_t len);

int main() {

    //
    // 1. Creating Socket
    //
    int s = socket(PF_INET, SOCK_RAW, 1);

    //
    //  -> Exit the app if the socket failed to be created
    //
    if(s <= 0)
    {
        perror("Socket Error");
        exit(0);
    }

    //
    // 2. Create the ICMP Struct Header
    //
    typedef struct {
        uint8_t type;
        uint8_t code;
        uint16_t chksum;
        uint16_t identifier;
        uint16_t sequence_number;
        uint32_t data;
    } icmp_hdr_t;

    //
    //  3. Use the newly created struct to make a variable.
    //
    icmp_hdr_t pckt;

    //
    //  4. Set the apropriate values to our struct, which is our ICMP header
    //
    pckt.type = 8;              // The echo request is 8
    pckt.code = 0;              // No need
    pckt.chksum = 0;            // The chacksum first needs to be calcualted
    pckt.identifier = getpid(); // Set a random Nr. in this case the app ID
    pckt.sequence_number = 1;   // Normaly you would increment this nummber
    pckt.data = 0;              // We don't send anything.

    // 
    //  5. Calculate the checksum based on the whole header, and only then
    //     you add it to the header.
    // 
    pckt.chksum = checksum((uint16_t *)&pckt, sizeof(pckt));

    //
    //  6. Creatign a IP Header from a struct that exists in another library
    //  
    struct sockaddr_in addr;
    addr.sin_family = AF_INET;
    addr.sin_port = 0;
    addr.sin_addr.s_addr = inet_addr("8.8.8.8");

    //
    //  7. Send our PING 
    //
    int actionSendResult = sendto(s, &pckt, sizeof(pckt), 
                                  0, (struct sockaddr*)&addr, sizeof(addr));

    //
    //  -> Exit the app if the option failed to be set
    //
    if(actionSendResult < 0)
    {
        perror("Ping Error");
        exit(0);
    }

    //
    //  8. Prepare all the necesary variable to handle the response
    //
    unsigned int resAddressSize;
    unsigned char res[30] = "";
    struct sockaddr resAddress;

    // 
    //  9. Creating the struct to better handle the response
    // 
    typedef struct {
        uint8_t type;
        uint8_t code;
        uint16_t checksum;
        uint16_t identifier;
        uint16_t sequence_number;
    } icmp_response_t;

    //
    //  10. Read the response from the remote host
    //
    int ressponse = recvfrom(s, res, sizeof(res), 0, &resAddress, 
                             &resAddressSize);

    //
    //  -> Display the response by accessign the struct
    //
    if(ressponse > 0)
    {
        // 
        //  11. Create the response variable usign our custom struct
        // 
        icmp_response_t* echo_response;

        //
        //  12. Map our resposne to our response struct starting from byte 20
        //
        echo_response = (icmp_response_t *)&res[20];

        // 
        //  -> Log the data that we'v got back
        // 
        printf(
            "type: %x, code: %x, checksum: %x, identifier: %x, sequence: %x\n",
            echo_response->type,
            echo_response->code,
            echo_response->checksum,
            echo_response->identifier,
            echo_response->sequence_number
        );

        exit(0);
    } 
    else 
    {
        perror("Response Error");
        exit(0);
    }

    return 0;
}

//
//  () Checksum Method
//
int32_t checksum(uint16_t *buf, int32_t len)
{
    //
    //  1. Variable needed for the calculation
    //
    int32_t nleft = len;    // Save how big is the header
    int32_t sum = 0;        // Container for the calcualted value
    uint16_t *w = buf;      // Save the first 2 bytes of the header
    uint16_t answer = 0;    // The state of our final anwser

    //
    //  2. Summ evry other byte from the header
    //
    while(nleft > 1)
    {
        sum += *w++;
        nleft -= 2;
    }

    //
    //  3. No idea
    //
    if(nleft == 1)
    {
        *(uint16_t *)(&answer) = *(uint8_t *)w;
        sum += answer;
    }

    //
    //  4. Needed conversions
    //
    sum = (sum >> 16) + (sum & 0xFFFF);
    sum += (sum >> 16);

    //
    //  5. Invert the bits
    //
    answer = ~sum;

    //
    //  -> Result
    //
    return answer;
}

File: NodeJS/pingWithChecksum.js
'use strict';

let raw = require ("raw-socket");
let header = [];

//
//  1. Create the structure where we are going to 
//     save our header
//
let type            = new Buffer(1);
let code            = new Buffer(1);
let chksum          = new Buffer(2);
let identifier      = new Buffer(2);
let sequence_number = new Buffer(2);
let data            = new Buffer(4);

//
//  2. Wrtie the apropriate values.
//
type.writeUInt8(0x8, 0);
code.writeUInt8(0x0, 0);
chksum.writeUInt16LE(0x0, 0);
identifier.writeUInt16LE(process.pid, 0);
sequence_number.writeUInt16LE(0x0, 0);
data.writeUInt32LE(0x0, 0);

//
//  3. push each separated buffer to the array
//
header.push(type);
header.push(code);
header.push(chksum);
header.push(identifier);
header.push(sequence_number);
header.push(data);

let headerConcat2 = new Buffer.concat(header);

let buf = new Buffer(2);
let int = checksum(header);
buf.writeUInt16LE(int);

header[2] = buf;

//
//  4. Combine all the buffers in to one
//
let headerConcat = new Buffer.concat(header);

//
//  5. Creatign the socket uign the ICMP protocol
//
var socket = raw.createSocket(
    {
        protocol: raw.Protocol.ICMP
    }
);

//
//  6. Sendign the request for a ping
//
socket.send(headerConcat, 0, 12, "8.8.8.8", function(error, bytes)
    {
        // 
        //  -> If there is any error, show it.
        // 
        if (error) 
        {
            console.log(error.toString());
        }
    }
);

//
//  7. Listent for the remote host response
//
socket.on("message", function (buffer, source) {


    // 
    //  8. Create a buffer that will hold just our ICMP reaply, we don't need 
    //     the whole TCP blob :)
    // 
    let icmpResponseBuffer = new Buffer(8);
    
    // 
    //  9.Copy only the fragment from the response that interest us,
    //    startign at byte 20
    // 
    buffer.copy(icmpResponseBuffer, 0, 20);

    // 
    //  10. Create all the buffers where we are goignto store the different
    //      information from the ICMP reply.
    // 
    let type = new Buffer(1);
    let code = new Buffer(1);
    let checksum = new Buffer(2);
    let identifier = new Buffer(2);
    let sequence_number = new Buffer(2);
    
    // 
    //  11. Copy bytes in to the apropierate buffer
    // 
    icmpResponseBuffer.copy(type, 0, 0);
    icmpResponseBuffer.copy(code, 0, 1);
    icmpResponseBuffer.copy(checksum, 0, 2);
    icmpResponseBuffer.copy(identifier, 0, 4);
    icmpResponseBuffer.copy(sequence_number, 0, 6);
    
    //
    //  -> Display in a human readable form the resposne that we got
    //
    console.log("type: %s, code: %s, checksum: %s, identifier: %s, sequence: %s",
                type.toString('hex'),
                code.toString('hex'),
                checksum.toString('hex'),
                identifier.toString('hex'),
                sequence_number.toString('hex')
            );

    //
    //  12. Once we have our response we can exit the app
    //
    process.exit()

});

//
//  () Checksum Method
//
function checksum(array)
{
    //
    //  1. Variable needed for the calculation
    //
    let buffer = new Buffer.concat(array);
    let twoByteContainer = new Buffer(2);
    let position = 0;
    let answer = 0;
    let sum = 0;
    
    //
    //  2. Read only the first two bytes from the buffor 
    //
    for(let i = 0; i < 6; i++)
    {
        //
        //  1. Copy two bytes from the original buffer
        //
        buffer.copy(twoByteContainer, 0, position);

        //
        //  2. read the buffer as unsigned integer in little endian mode
        //
        let decimal = twoByteContainer.readUIntLE(0, 2);

        //
        //  3. Increse the position by two
        //
        position += 2
    
        //
        //  4. Sum the new value with the previous one
        //
        sum += decimal;
    }

    //
    //  3. Needed conversions
    //
    sum = (sum >> 16) + (sum & 0xFFFF);
    sum += (sum >> 16);
    
    //
    //  4. Invert the bits
    //    
    answer = ~sum;
    
    //
    //  5. Interprete the number as Unsigned
    //    
    let uintResult = (new Uint16Array([answer]))[0];

    //
    //  -> Result
    //
    return uintResult;
}
